### PR TITLE
 support rosa and hypershift clusters

### PR DESCRIPTION
**Pull request by myakove**
Install IPI cluster on AWS:
```
poetry run python app/cli.py -a create --registry-config-file=registry-config.json --clusters-install-data-directory=/tmp/ocp-clusters --s3-bucket-name=openshift-cli-installer --s3-bucket-path=install-folders  -c 'name=ipi1;base_domain=aws.interop.ccitredhat.com;platform=aws;region=us-east-2;version=4.14.0-ec.2;worker_flavor=m5.xlarge'
```

Install Rosa cluster:
```
poetry run python app/cli.py -a create --ocm-token=$OCM_TOKEN --ocm-env=stage --clusters-install-data-directory=/tmp/ocp-clusters -c 'name=rosa-managed1;platform=rosa;region=us-east-2;version=4.13.4;compute-machine-type=m5.xlarge;replicas=2;channel-group=candidate;expiration-time=4;timeout=1h' 
```

Install Hypershift cluster:
```
poetry run python app/cli.py -a create --ocm-token=$OCM_TOKEN --ocm-env=stage --clusters-install-data-directory=/tmp/ocp-clusters -c 'name=hypershift1;platform=hypershift;region=us-west-2;version=4.13.4;compute-machine-type=m5.xlarge;replicas=2;channel-group=candidate;expiration-time=4;timeout=1h'
```

### CODE REVIEW

File: app/libs/rosa_clusters.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+                "expiration-time"
+            ] = f"{(datetime.now() + timedelta(hours=expiration_time)).isoformat()}Z"
+
+    return clusters
+
+
+def rosa_create_cluster(cluster_data):
+    hosted_cp_arg = "--hosted-cp"
+    ignore_keys = ("name", "platform", "ocm-env", "ocm-token")
+    command = "create custer --sts "
```
Comment (omrirh) [inline app/libs/rosa_clusters.py:27]:
```suggestion
    command = "create cluster --sts "
```

**Review by omrirh — APPROVED** [review]

**Comment by myakove** [comment]
pre-commit.ci run

File: app/cli.py [hunk 2]
```diff
@@ -65,96 +49,27 @@ def verify_processes_passed(processes, action):
...
 
-
-def create_openshift_cluster(cluster_data, s3_bucket_name, s3_bucket_path):
-    create_or_destroy_cluster(
-        cluster_data=cluster_data,
-        action="create",
-        s3_bucket_name=s3_bucket_name,
-        s3_bucket_path=s3_bucket_path,
-    )
+    elif cluster_platform in (AWS_MANAGED_STR, HYPERSHIFT_STR):
```
Comment (rnetser) [inline app/cli.py:62]:
```suggestion
    elif cluster_platform in (ROSA_MANAGED_STR, HYPERSHIFT_STR):
```

File: app/cli.py [hunk 3]
```diff
@@ -204,6 +125,18 @@ def get_local_ssh_key():
     help="S3 bucket path to store the backups",
     show_default=True,
 )
+@click.option(
+    "--ocm-env",
+    help="OCM env to log in into. needed for managed AWS cluster",
```
Comment (rnetser) [inline app/cli.py:130]:
```suggestion
    help="OCM env to log in into. Needed for managed AWS cluster",
```

File: app/cli.py [hunk 4]
```diff
@@ -216,7 +149,7 @@ def get_local_ssh_key():
 Required parameters:
     name: Cluster name.
     base_domain: Base domain for the cluster.
-    platform: Cloud platform to install the cluster on. (Currently only AWS supported).
+    platform: Cloud platform to install the cluster on. (Currently only AWS IPI supported).
```
Comment (rnetser) [inline app/cli.py:152]:
```suggestion
    platform: Cloud platform to install the cluster on. (Currently supported:  AWS IPI, ROSA and Hypershift.
```

File: app/cli.py [hunk 5]
```diff
@@ -231,41 +164,63 @@ def get_local_ssh_key():
...
+        _cluster for _cluster in cluster if _cluster["platform"] == AWS_STR
+    ]
+    aws_managed_clusters = [
+        _cluster
+        for _cluster in cluster
+        if _cluster["platform"] in (AWS_MANAGED_STR, HYPERSHIFT_STR)
+    ]
+    if aws_ipi_clusters or aws_managed_clusters:
+        for _cluster in aws_ipi_clusters + aws_managed_clusters:
+            set_and_verify_aws_credentials(region_name=_cluster["region"])
```
Comment (rnetser) [inline app/cli.py:192]: extract the region so we do not have to run muliple times on the same region

File: app/libs/rosa_clusters.py [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+        ocm_client=ocm_client,
+    )
+
+    res = rosa.cli.execute(
+        command=f"list oidc-config --region={aws_region}",
+        aws_region=aws_region,
+        ocm_client=ocm_client,
+    )["out"]
+
+    _oidc_config_id = [
```
Comment (rnetser) [inline app/libs/rosa_clusters.py:86]:
```suggestion
    cluster_data["oidc-config-id"] = [
```

File: app/libs/rosa_clusters.py [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+    cluster_object.wait_for_cluster_ready(wait_timeout=cluster_data["timeout"])
+
+    auth_path = os.path.join(cluster_data["install-dir"], "auth")
+    Path(auth_path).mkdir(parents=True, exist_ok=True)
+
+    with open(os.path.join(auth_path, "kubeconfig"), "w") as fd:
+        fd.write(yaml.dump(cluster_object.kubeconfig))
+
+    with open(os.path.join(auth_path, "kubeadmin-password"), "w") as fd:
+        fd.write(cluster_object.kubeadmin_password)
```
Comment (rnetser) [inline app/libs/rosa_clusters.py:243]: different func?

File: app/libs/rosa_clusters.py [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+    job = Job(
+        client=cluster_object.ocp_client,
+        name="osd-cluster-ready",
+        namespace="openshift-monitoring",
+    )
+    job.wait_for_condition(
+        condition=job.Condition.COMPLETE,
+        status="True",
+        timeout=tts(ts="40m"),
+    )
```
Comment (rnetser) [inline app/libs/rosa_clusters.py:254]: different func?

**Review by rnetser — CHANGES_REQUESTED** [review]

File: app/libs/rosa_clusters.py (hunk 7 above)
Comment (myakove) [reply app/libs/rosa_clusters.py:258]: why?

Comment (rnetser) [reply app/libs/rosa_clusters.py:248]: to have one place to do it and make it easier to follow the function

Comment (myakove) [reply app/libs/rosa_clusters.py]: Done

**Review by rnetser — APPROVED** [review]

**Pull request merged**